### ISSUE TITLE
How to free memory after query

### ISSUE THREAD

**Issue by ghost** [libspatialindex/libspatialindex#100]
Hello,
i have discovered a strange issue thats keeping me from using the library the way i want. I have implemented a custom query strategy implementing my own class derived from IQueryStrategy and using the queryStrategy method to search the tree. However, after the extensive search, memory is still alocated inside the heap and i have spent 2 days trying to figure out where is the memory allocated and why it is not freed when i call delete on the rtree class with no success. The memory only gets released when the process ends, but i need to do this inside a daemon process. The queryStrategy function is calling internally readNode, after that inside function i am doing nothing else but requesting nodes from tree until i stop it with setting up hasNext variable to false. What can be the problem?

**Comment by brano543** [comment]
I have written myself a test case where i compare In-memory and On-disk created r-tree and query. It doesn't matter which one i use, there are still resources on heap allocated which are only released when the program terminates. With real dataset it is a few hundreds of MB allocated on heap and never freed when doing extensive search(500k nodes traversed in tree). Even on random generated tree with 100k points and using in-memory implementation, output of my program is 
Allocated memory(except files) is 7.33594 MB (tree before query)
Found 100000 results
Allocated memory(except files) is 7.52344 MB (tree after query and delete on tree and memory_manage)

I guess this has something to do with the member variables inside Rtree.h and this issue might have been referenced in #60 
Tools::PointerPool<Point> m_pointPool;
Tools::PointerPool<Region> m_regionPool;
Tools::PointerPool<Node> m_indexPool;
Tools::PointerPool<Node> m_leafPool;
Correct me if i am wrong, but the purpose of PointerPool is to get already stored pointer or create a new pointer. I think the issue might be in PoolPointer,because this class tracks how many references there are and only when there is one last it clears the pointer. Maybe inside the code it is called one time less and that means that only when all destructors are called at the end of process it is actually freed, because it cant correctly track the count of references, but i am not sure about this. What is your opinion on this?

**Comment by brano543** [comment]
I am attaching my simple test program if any good soul decides to help me. First class just reads /proc/pid/status and parses the memory usage. Problem is when i decide to store the data inside a vector, even after the destructor is called,the data is still in memory. I don`t understand why. I have even tried to replace vector of tuple with vector of pointers but still had the same issue. I even tried standard array and deleting,no matter what i did it still sits inside memory. 
```

#include <spatialindex/SpatialIndex.h>
#include <cstring>

class MemoryUsage
{
    //VmRSS is the measure of how much psysical RAM the process is actually using (allocated and used)
    double m_vmrss = 0;

    //Part of the RSS is allocated in real memory blocks (other than mapped doubleo a file or device).
    //This is anonymous memory (anon-rss) and there is also RSS memory blocks that are mapped doubleo devices and files (file-rss)
    double m_rssanon = 0;
    double m_rssfile = 0;
    double m_rssshmem = 0; //shared memory

    static double parse_number(const std::string &line)
    {
        const auto f = line.find_first_of("0123456789");
        const auto l = line.find_last_of("0123456789");
        return std::atoi(line.substr(f, l - f + 1).c_str());
    }

  public:
    /**
         * Get the memory usage for the current process. The constructor will
         * get the memory usage
         */
    MemoryUsage()
    {
        monitor_usage();
    }

    //this function checks /proc/self/status, parses the values and updates member values if current value is higher
    void monitor_usage()
    {
        static const char *filename = "/proc/self/status";
        std::ifstream status_file(filename);

        if (status_file.is_open())
        {
            std::string line;
            while (!status_file.eof())
            {
                std::getline(status_file, line);
                if (line.substr(0, 5) == "VmRSS")
                {
                    double current_vmrss = parse_number(line);
                    if (current_vmrss > m_vmrss)
                    {
                        m_vmrss = current_vmrss;
                    }
                }
                if (line.substr(0, 7) == "RssAnon")
                {
                    double current_rssanon = parse_number(line);
                    if (current_rssanon > m_rssanon)
                    {
                        m_rssanon = current_rssanon;
                    }
                }
                if (line.substr(0, 7) == "RssFile")
                {
                    double current_rssfile = parse_number(line);
                    if (current_rssfile > m_rssfile)
                    {
                        m_rssfile = current_rssfile;
                    }
                }
                if (line.substr(0, 8) == "RssShmem")
                {
                    double current_rssshmem = parse_number(line);
                    if (current_rssshmem > m_rssshmem)
                    {
                        m_rssshmem = current_rssshmem;
                    }
                }
            }
        }
    }

    // Return total memory usage in MBytes
    double vm_rss() const
    {
        return m_vmrss / 1024;
    }

    // Return memory usage not mapped to files in MBytes
    double rss_anon() const
    {
        return m_rssanon / 1024;
    }

    // Return memory usage for files and devices mapping in MBytes
    double rss_file() const
    {
        return m_rssfile / 1024;
    }

    //Return memory usage for shared memory in MBytes
    double rss_shared() const
    {
        return m_rssshmem / 1024;
    }
};

struct RTreeData
{
    uint32_t source;
};

using namespace std;
namespace si = SpatialIndex;

int main()
{
    {
        //create in memory r-tree
        si::IStorageManager *memory_manager = si::StorageManager::createNewMemoryStorageManager();
        double const fill_factor = 0.7;      //default 0.7 (R*), 0.0-0.5 for linear/quadratic
        uint32_t const index_capacity = 100; //default 100
        uint32_t const leaf_capacity = 100;  // default: 100
        uint32_t const dimension = 2;
        si::RTree::RTreeVariant variant = si::RTree::RV_RSTAR;
        si::id_type index_id;
        si::ISpatialIndex *rtree = si::RTree::createNewRTree(*memory_manager, fill_factor, index_capacity, leaf_capacity, dimension, variant, index_id);

        //fill up tree with 10k test points
        {
            vector<pair<double, double>> points;

            //generate random points
            double lat = 40.0;
            double lon = 17.0;
            for (int i = 0; i < 10000; ++i)
            {
                points.push_back(make_pair(lat, lon));
                lat += 0.0001;
                lon += 0.0001;
            }

            uint32_t i = 1;
            for (auto &p : points)
            {
                double plow[] = {get<0>(p), get<1>(p)};
                double phigh[] = {get<0>(p), get<1>(p)};
                si::Region mbr(plow, phigh, 2);
                si::id_type item_id(i);
                RTreeData data{i};
                rtree->insertData(sizeof(data), reinterpret_cast<byte *>(&data), mbr, item_id);
                i++;
            }
        }

        MemoryUsage ms;
        cout << "Allocated memory(except files) is " << ms.rss_anon() << " MB" << endl;

        //try to query all 10k points
        {
            TestPoiVisitor nearest_neighbor_visitor;
            double q_coords[] = {48.145584, 17.1271079};
            si::Point q(q_coords, 2);
            uint32_t k = 10000;
            rtree->nearestNeighborQuery(k, q, nearest_neighbor_visitor);
            vector<tuple<si::Region, uint32_t>> nearestNeighborCandidates = nearest_neighbor_visitor.getNearestCandidates();
            cout << "Found " << to_string(nearestNeighborCandidates.size()) << " results" << endl;
        }

        delete rtree;
        delete memory_manager;

        //delete test file
        remove("test_lmdb.bin");

        ms.monitor_usage();
        cout << "Allocated memory(except files) is " << ms.rss_anon() << " MB" << endl;
    }
return 0;
}
```

**Comment by hobu** [comment]
I can sympathize with the frustration and the only advice I can offer is to look to the shared pointer implementation in Tools and get intimate with valgrind. It seems like you have the tools to replicate and get close, at least.

**Comment by brano543** [comment]
Thank you sir, but there is also something different else going on here. I am not programming in C++ for so long, it´s actually only a 3 years, however there is one thing i never anticipated and recently found out. I was always taught that every variable gets destroyed after a scope ends if its not dynamically allocated memory with new. However i have tried to cut this problem to smaller pieces and i was really surprised. I have measured memory usage as in above by reading /proc/self/status and VMRSS property which is actually physical memory the process uses. I got into this smallest piece of code. Simple vector with different classes or variables. This is what i found out:
//lets say memory usage 0 MB
{
vector<string> test;
//lets call 1000 time push some arbitrary char;
for (int i=0; i <1000; ++i) test.push_back("random string")
}
}
//what would i expect that physical memory usage is 0 MB here,because it is deleted
//but here we get number like 3 MB

And i wonder why is that so, when process ends everything is OK, but here we are talking about daemon process, we want to release the memory. So there is no mechanism, at least it doesn´t work it seems for classes/variables with dynamically allocated size, so the memory is not released back to OS, so i just run out of memory this way after few requests. Maybe this has nothing to do with the library, so in the next few days i will try to replace in worker threads every vector and use New and just free this after the thread ends. If that wont work, i have read on Stackoverflow here, that maybe some solution like Jemalloc might help. Anyways, i will update this thread when i learn something. To me it seems C++ acts like C while working with automatic memory management for dynamically allocated variables at least.

So in summary in my simple test.
vector<tuple<int, uint>> test; worked just as i expect it to when i pushed something inside and went out of scope

**Comment by kounoupis** [comment]
Typically, libc does not do any effort to return freed memory back to the
system. Memory is reclaimed when the process terminates. If you are running
a deamon that needs to allocate a lot of temporary memory, you need to fork
a new process for each request.
